### PR TITLE
Combining baked and dynamic shadows nicely (lightmap)

### PR DESCRIPTION
**Pull request by BitOfGold**
https://www.bitofgold.com/lightmaptest/
A little demonstration to show what is wrong with the previous ligthmap calculation - no ambient lighting, no other light, no specular highlights.

Does not break the previous lightmap calculation.

material.lightmapTexture needs to be set to activate light map.

light.lightmapExcluded
if true, gets de diffuse lighting from the light map. adds specular highlights mixed widh baked and dynamic shadows.

if false, lightmap calculation is not affected. ("Old" lightmap, material.useLightmapAsShadowmap works as before (simply multiplying color.rgb after composition). Does not look right.)

light.lightmapNoSpecular
if true, does not even calculate the light functions, only the baked lightmap + dynamic shadow. (no specular then from this light, only the baked diffuse)

### CODE REVIEW

**Comment by BitOfGold** [comment]
I commited to the wrong branch I think, mixed this with my previus pull request.
It's OK now.

**Comment by deltakosh** [comment]
This is really cool
I would not use a double boolean as if you do not set ligtmapExcluded then noSpecular as no effect.

I would prefer some kind of mode enum:
* None
* LightmapExcluded
* LightmapExcludedNoSpecular

**Comment by BitOfGold** [comment]
Thanks!
I'm on it.

**Comment by BitOfGold** [comment]
Oh, working on two pull requests on two branches is too much for me...
Now **both looks fine**, but i do not now how to update these branches with the updates that correct the following error:

`babylon.js:23 Uncaught ReferenceError: SIMD is not defined`
It was an error that is now corrected. My 'master' branch is ok

I'm not an experienced github user, and I pull my hair out. If I update the 'lightmapsexcluded' branch from the 'master', why that error stays, I cannot comprehend... I see no other option than deleting my fork and starting over.
Sorry!

I hope theese 2 PG-s get in soon.

**Comment by deltakosh** [comment]
No rush:)

File: src/Materials/babylon.materialHelper.ts [hunk 1]
```diff
@@ -103,6 +104,20 @@
                     }
                 }
 
+                if (light.lightmapMode != 0 ) {
```
Comment (deltakosh) [inline src/Materials/babylon.materialHelper.ts]: to ensure people know which value to use, I suggest also creating named values like here:
https://github.com/BabylonJS/Babylon.js/blob/master/src/babylon.engine.ts#L199

Comment (deltakosh) [inline src/Materials/babylon.materialHelper.ts]: And this is where the naming will be important to help users understand each mode

**Comment by deltakosh** [comment]
Good job!

**Comment by nockawa** [comment]
great job! it deserves credits in the what's new, right?

**Comment by BitOfGold** [comment]
I refreshed my little demonstration app:
[https://www.bitofgold.com/lightmaptest/](https://www.bitofgold.com/lightmaptest/)

**Comment by deltakosh** [comment]
Excellent

**Review by GuDuJian-J-Zhang — COMMENTED** [review]
have a question about this PR:
why lightmode is a pre light property
since we can specify only one lightmap:

// Create lightmap texture on a material
    var lightmap = new BABYLON.Texture("textures/candleopacity.png", scene);
    var groundMaterial = new BABYLON.StandardMaterial("groundMaterial", scene);
    groundMaterial.lightmapTexture = lightmap;

light.lightmapMode = BABYLON.Light.LIGHTMAP_SPECULAR;

**Comment by BitOfGold** [comment]
It is per-light, because the property decides the behaviour of the light, not the material:

- default (calculates the light normally)
- specular (only calculates the specular lighting, this light is pre-baked with materials having a light map)
- shadows only (no light calculation on materials with a light map)

**Comment by GuDuJian-J-Zhang** [comment]
> It is per-light, because the property decides the behaviour of the light, not the material:
> 
> * default (calculates the light normally)
> * specular (only calculates the specular lighting, this light is pre-baked with materials having a light map)
> * shadows only (no light calculation on materials with a light map)

thanks for your response. but there are two more questions if there are multiple lights in a scene.

1. how to specify multiple lightmaps for a material ?

2. how can i decide which light will be influenced by which lightmap?

**Pull request merged**